resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/K8sSubmitOps.scala

```diff
@@ -0,0 +1,192 @@
+  override def executeOnPod(pName: String, namespace: Option[String], sparkConf: SparkConf)
+      (implicit client: KubernetesClient): Unit = {
+    val gracePeriodSeconds = getGracePeriod(sparkConf)
+    val podToDelete = {
+     namespace match {
+       case Some(ns) => client.pods.inNamespace(ns)
+       case None => client.pods
+     }
+    }.withName(pName)
+
+    gracePeriodSeconds match {
+      case Some(period) => podToDelete.withGracePeriod(period).delete()
+      case _ => podToDelete.delete()
+    }
+  }
+
+  override def executeOnGlob(pods: List[Pod], namespace: Option[String], sparkConf: SparkConf)
+      (implicit client: KubernetesClient): Unit = {
+    if (pods.nonEmpty) {
+      pods.foreach { pod => printMessage(s"Deleting driver pod: ${pod.getMetadata.getName}.") }
+      val nClient = {
+        namespace match {
+          case Some(ns) => client.pods.inNamespace(ns)
+          case None => client.pods
+        }
+      }
+
+      val gracePeriodSeconds = getGracePeriod(sparkConf)
+      gracePeriodSeconds match {
+        case Some(period) =>
+          // this is not using the batch api because no option is provided
+          // when using the grace period.
+          pods.foreach { pod =>
+             nClient
+               .withName(pod.getMetadata.getName)
+               .withGracePeriod(period)
+               .delete()
+          }
+        case _ => nClient.delete(pods.asJava)
+      }
+    } else {
+      printMessage("No applications found.")
+    }
+  }
+}
+
+private class ListStatus extends K8sSubmitOp {
+  override def executeOnPod(pName: String, namespace: Option[String], sparkConf: SparkConf)
+      (implicit client: KubernetesClient): Unit = {
+    val pod = {
+      namespace match {
+        case Some(ns) => client.pods.inNamespace(ns)
+        case None => client.pods
+      }
+    }.withName(pName)
+      .get()
+
+    printMessage("Application status (driver): " +
+      Option(pod).map(formatPodState).getOrElse("unknown."))
+  }
+
+  override def executeOnGlob(pods: List[Pod], ns: Option[String], sparkConf: SparkConf)
+      (implicit client: KubernetesClient): Unit = {
+    if (pods.nonEmpty) {
+      for (pod <- pods) {
+        printMessage("Application status (driver): " +
+          Option(pod).map(formatPodState).getOrElse("unknown."))
+      }
+    } else {
+      printMessage("No applications found.")
+    }
+  }
+}
+
+private[spark] class K8SSparkSubmitOperation extends SparkSubmitOperation
+  with CommandLineLoggingUtils {
+
+  private def isGlob(name: String): Boolean = {
+    name.last == '*'
+  }
+
+  def execute(submissionId: String, sparkConf: SparkConf, op: K8sSubmitOp): Unit = {
+    val master = KubernetesUtils.parseMasterUrl(sparkConf.get("spark.master"))
+    val sIdParts = submissionId.split(":") match {
+      case Array(ns, pName) => Some((Some(ns), pName))
+      case Array(pName) => Some((None, pName))
+      case _ => None
```

Comment:
Shouldn't this be an error?

Or maybe you should use `.split(":", 2)`?

(You seem to handle `None` below, but why wait until then?)

Reply:
I like this pattern matching. I could handle it there and leave the other matching latter empty. Anyway I can update it.